File: js/src/stores/redisStore.js
/**
 * admitiq/stores/redisStore — optional Redis-backed revocation store.
 *
 * A ready-made, race-condition-safe "has this token been used" store for
 * anyone running multiple servers/processes checking the same tokens,
 * without needing the hosted AdmitiQ API. Still fully self-hosted and free.
 *
 * Requires the 'redis' package (node-redis v4+):
 *   npm install redis
 *
 * Usage:
 *   const client = redis.createClient({ url: 'redis://localhost:6379' });
 *   await client.connect();
 *   const store = new RedisRevocationStore({ client });
 *   const payload = await verify(token, SECRET, store.isRevoked.bind(store));
 *   // only after you've decided to actually honor this scan:
 *   await store.markUsed(payload.jti);
 */
class RedisRevocationStore {
  /**
   * @param {object} options
   * @param {object} options.client - an already-connected node-redis v4 client
   * @param {number} [options.ttlSeconds=86400] - how long to remember a used jti
   * @param {string} [options.prefix="admitiq:used:"]
   */
  constructor({ client, ttlSeconds = 86400, prefix = "admitiq:used:" } = {}) {
    if (!client) {
      throw new Error(
        "RedisRevocationStore requires an already-connected 'redis' client. " +
          "Install with: npm install redis — see README for setup."
      );
    }
    this.client = client;
    this.ttlSeconds = ttlSeconds;
    this.prefix = prefix;
  }

  /**
   * Atomically mark a token as used. Returns true if this was the FIRST time
   * (the scan should be allowed), false if already used (block as reuse).
   */
  async markUsed(jti) {
    const key = `${this.prefix}${jti}`;
    const result = await this.client.set(key, "1", { NX: true, EX: this.ttlSeconds });
    return result === "OK";
  }

  /** Pass this bound method directly as the isRevoked argument to verify(). */
  async isRevoked(jti) {
    const key = `${this.prefix}${jti}`;
    const exists = await this.client.exists(key);
    return exists > 0;
  }
}

module.exports = { RedisRevocationStore };

File: python/admitiq/stores/redis_store.py
"""
admitiq.stores.redis_store — optional Redis-backed revocation store.

A ready-made, race-condition-safe "has this token been used" store for
anyone running multiple servers/processes checking the same tokens, without
needing the hosted AdmitiQ API. Still fully self-hosted and free.

Requires the 'redis' package:
    pip install admitiq[redis]
"""
from typing import Optional


class RedisRevocationStore:
    """
    Usage:
        store = RedisRevocationStore(url="redis://localhost:6379/0")
        payload = verify(token, secret=SECRET, is_revoked=store.is_revoked)
        # only after you've decided to actually honor this scan:
        store.mark_used(payload["jti"])
    """

    def __init__(
        self,
        redis_client=None,
        url: Optional[str] = None,
        ttl_seconds: int = 86400,
        prefix: str = "admitiq:used:",
    ):
        """
        Args:
            redis_client: an existing redis.Redis instance (preferred if you already have one)
            url: alternatively, a redis:// URL to connect with
            ttl_seconds: how long to remember a used jti — should be >= your longest token TTL
            prefix: key prefix used in Redis
        """
        if redis_client is not None:
            self._client = redis_client
        else:
            try:
                import redis
            except ImportError as e:
                raise ImportError(
                    "RedisRevocationStore requires the 'redis' package. "
                    "Install it with: pip install admitiq[redis]"
                ) from e
            self._client = redis.Redis.from_url(url or "redis://localhost:6379/0")
        self._ttl = ttl_seconds
        self._prefix = prefix

    def mark_used(self, jti: str) -> bool:
        """
        Atomically mark a token as used. Returns True if this was the FIRST
        time it was marked (the scan should be allowed), False if it was
        already used (this is a reuse attempt and should be blocked).
        """
        key = f"{self._prefix}{jti}"
        was_set = self._client.set(key, "1", nx=True, ex=self._ttl)
        return bool(was_set)

    def is_revoked(self, jti: str) -> bool:
        """Pass this method directly as the `is_revoked` argument to verify()."""
        key = f"{self._prefix}{jti}"
        return bool(self._client.exists(key))
